lab-nassir/test/cafe-route-test.js

```diff
@@ -0,0 +1,385 @@
+        request.get(`${url}/api/cafe/34`)
+        .end((err, res) => {
+          expect(res.status).to.equal(404);
+          expect(res.text).to.equal('NotFoundError');
+          done();
+        });
+      });
+    });
+
+    describe('Testing GET /api/cafe with NO ID', function(){
+
+      before(done => {
+        new Cafe(exampleCafe).save()
+        .then(new Cafe(exampleCafeTwo).save())
+        .then(() => done())
+        .catch(done);
+        return;
+      });
+
+      after(done => {
+        if(this.tempCafe) {
+          Cafe.remove({})
+          .then(() => done())
+          .catch(done);
+          return;
+        }
+        done();
+      });
+
+      it('Should return a status code of 200 and an array of docs', done => {
+        request.get(`${url}/api/cafe/`)
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(200);
+          expect(res.body instanceof Array).to.equal(true);
+          expect(res.body.length).to.equal(2);
+          expect(res.body[0].name).to.equal('Test Cafe');
+          expect(res.body[1].name).to.equal('Test Cafe Two');
+          done();
+          this.tempCafe = res.body[0];
+        });
+      });
+    });
+
+  });
+
+  describe('Testing PUT /api/cafe', function(){
+
+    describe('Testing PUT /api/cafe with VALID BODY', function(){
+
+      before(done => {
+        new Cafe(exampleCafe).save()
+        .then(cafe => {
+          this.tempCafe = cafe;
+          done();
+        })
+        .catch(done);
+        return;
+      });
+
+      after(done => {
+        if(this.tempCafe) {
+          Cafe.remove({})
+          .then(() => done())
+          .catch(done);
+          return;
+        }
+        done();
+      });
+
+      it('Should return status code 200 and a valid body', done => {
+        request.put(`${url}/api/cafe/${this.tempCafe._id}`)
+        .set('Content-type', 'application/json')
+        .send({name: 'New Test Name', address: 'New Test Address'})
+        .end((err, res) => {
+          if (err) return done(err);
+          expect(res.status).to.equal(200);
+          expect(res.body.name).to.equal('New Test Name');
+          expect(res.body.address).to.equal('New Test Address');
+          let testTimestamp = new Date(res.body.timestamp);
+          expect(testTimestamp.toString()).to.equal(exampleCafe.timestamp.toString());
+          done();
+        });
+      });
+    });
+
+    describe('Testing PUT /api/cafe with INVALID BODY', function(){
+
+      before(done => {
+        new Cafe(exampleCafe).save()
+        .then(cafe => {
+          this.tempCafe = cafe;
+          done();
+        })
+        .catch(done);
+        return;
+      });
+
+      after(done => {
+        if(this.tempCafe) {
+          Cafe.remove({})
+          .then(() => done())
+          .catch(done);
+          return;
+        }
+        done();
+      });
```

Comment:
Same as above -- you can set up your testing environment in the outer describe block for all PUTs, valid and invalid.